pkg/controller/installation/windows/calico_windows_upgrader.go

```diff
@@ -0,0 +1,405 @@
+			// If the node has the latest version and it still has the upgrade
+			// script label, then it has just finished upgrading and needs the label
+			// removed. If it does not have the label, it's already upgraded and
+			// nothing needs to be done.
+			if node.Labels[common.CalicoWindowsUpgradeLabel] == common.CalicoWindowsUpgradeLabelInProgress {
+				windowsLog.V(1).Info(fmt.Sprintf("Node %v still has the upgrade label", node.Name))
+				inprogress[node.Name] = node
+			} else {
+				insync[node.Name] = node
+			}
+		}
+	}
+
+	windowsLog.V(1).Info(fmt.Sprintf("pending=%v, in-progress=%v, in-sync=%v", len(pending), len(inprogress), len(insync)))
+	return nil, pending, inprogress, insync
+}
+
+func (w *calicoWindowsUpgrader) upgradeCompleted(node *corev1.Node) bool {
+	_, variant, version := common.GetNodeVariantAndVersion(node)
+	if variant == w.install.Variant && version == w.getExpectedVersion() {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) isUpgradeFromCalicoToEnterprise(node *corev1.Node) bool {
+	_, variant, _ := common.GetNodeVariantAndVersion(node)
+	if variant == operatorv1.Calico && w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return true
+	}
+	return false
+}
+
+func (w *calicoWindowsUpgrader) getExpectedVersion() string {
+	if w.install.Variant == operatorv1.TigeraSecureEnterprise {
+		return components.EnterpriseRelease
+	}
+	return components.CalicoRelease
+}
+
+func (w *calicoWindowsUpgrader) updateWindowsNodes() {
+	if w.install == nil {
+		windowsLog.V(1).Info("Skipping update since installation not set yet")
+		return
+	}
+
+	err, pending, inprogress, insync := w.getNodeUpgradeStatus()
+	if err != nil {
+		windowsLog.Error(err, "Failed to get Windows nodes upgrade status")
+		return
+	}
+
+	for _, node := range inprogress {
+		if w.upgradeCompleted(node) {
+			if err := w.finishUpgrade(context.Background(), node); err != nil {
+				// Log the error and continue. We will retry when we update nodes again.
+				windowsLog.Info(fmt.Sprintf("Could not complete upgrade on node %v: %v", node.Name, err))
+				continue
+			}
+			// Successfully finished completing the upgrade. Moving the node
+			// from in-progress to in-sync.
+			insync[node.Name] = node
+			delete(inprogress, node.Name)
+		}
+	}
+
+	// Get the total # of windows nodes we can have upgrading using the
+	// maxUnavailable value, if the node upgrade strategy was respected.
+	numWindowsNodes := len(pending) + len(inprogress) + len(insync)
+	maxUnavailable, err := intstr.GetValueFromIntOrPercent(w.install.NodeUpdateStrategy.RollingUpdate.MaxUnavailable, numWindowsNodes, false)
+	if err != nil {
+		windowsLog.Error(err, "Invalid maxUnavailable value, falling back to default of 1")
+		maxUnavailable = defaultMaxUnavailable
+	}
+
+	for _, node := range pending {
+		// For upgrades from Calico -> Enterprise, we always upgrade regardless
+		// of maxUnavailable. For other upgrades, check that we have room
```

Comment:
For OS -> EE we want to upgrade nodes as soon as possible to limit the time during which there may be EE policy that isn't enforced on the OS node. On some platforms, depending on the node upgrade strategy in place, there could be a large fraction of windows nodes suddenly being upgraded.

Reply:
I don't think I agree - we're going through a lot of effort to ensure this is a rolling update. If we just wanted to upgrade these as fast as possible, then we wouldn't be implementing this complex logic in the operator. The purpose of rolling update is to prevent rolling out a bad config, and also to ensure things like BGP graceful restart work properly  - if we ignore max unavailable, we don't get those.

Now, if you wanted to limit this _only_ to new nodes that come up after enterprise has been installed, that might be a different story. But certainly taking a cluster with 20 Windows nodes running Calico and turning those into 20 enterprise nodes should be done in a rolling fashion.

Reply:
Discussed offline - going to leave this for now and fix in a future iteration. Definitely a bug, but one we're going to live with since the fix for the alternative is complex.